File: LinkDemo-ObjC/LinkDemo-ObjC/ViewControllers/EmbeddedSearch/PlaidEmbeddedLinkViewController.m
//
//  PlaidEmbeddedLinkViewController.m
//  LinkDemo-ObjC
//
//  Copyright © 2026 Plaid Inc. All rights reserved.
//

#import "PlaidEmbeddedLinkViewController.h"
#import "LinkKitObjectiveC.h"
@import LinkKitObjC;

typedef NS_ENUM(NSInteger, LoadingState) {
    LoadingStateLoading,
    LoadingStateReady,
    LoadingStateError
};

@interface PlaidEmbeddedLinkViewController ()

@property (nonatomic, copy) NSString *linkToken;
@property (nonatomic, assign) LoadingState state;
@property (nonatomic, copy, nullable) NSString *errorMessage;
@property (nonatomic, strong, nullable) UIView *embeddedSearchView;

@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *versionLabel;
@property (nonatomic, strong) UILabel *errorLabel;
@property (nonatomic, strong) UIView *embeddedViewContainer;

@end

@implementation PlaidEmbeddedLinkViewController

- (instancetype)initWithLinkToken:(NSString *)linkToken {
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        _linkToken = [linkToken copy];
        _state = LoadingStateLoading;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"Plaid Embedded Link";
    self.view.backgroundColor = [UIColor systemBackgroundColor];
    
    [self setupUI];
    [self createEmbeddedView];
}

- (void)setupUI {
    UIScrollView *scrollView = [[UIScrollView alloc] init];
    scrollView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:scrollView];
    
    UIStackView *stackView = [[UIStackView alloc] init];
    stackView.axis = UILayoutConstraintAxisVertical;
    stackView.spacing = 16;
    stackView.translatesAutoresizingMaskIntoConstraints = NO;
    [scrollView addSubview:stackView];

    self.titleLabel = [[UILabel alloc] init];
    self.titleLabel.text = @"Plaid Embedded Link View Example";
    self.titleLabel.font = [UIFont boldSystemFontOfSize:20];
    self.titleLabel.textAlignment = NSTextAlignmentCenter;
    self.titleLabel.numberOfLines = 0;
    [stackView addArrangedSubview:self.titleLabel];

    self.versionLabel = [[UILabel alloc] init];
    NSString *version = [PLKPlaid sdkVersion];
    self.versionLabel.text = [NSString stringWithFormat:@"LinkKit v%@", version];
    self.versionLabel.font = [UIFont systemFontOfSize:14];
    self.versionLabel.textColor = [UIColor secondaryLabelColor];
    self.versionLabel.textAlignment = NSTextAlignmentCenter;
    [stackView addArrangedSubview:self.versionLabel];

    self.errorLabel = [[UILabel alloc] init];
    self.errorLabel.font = [UIFont systemFontOfSize:14];
    self.errorLabel.textColor = [UIColor systemRedColor];
    self.errorLabel.textAlignment = NSTextAlignmentCenter;
    self.errorLabel.numberOfLines = 0;
    self.errorLabel.hidden = YES;
    [stackView addArrangedSubview:self.errorLabel];

    self.embeddedViewContainer = [[UIView alloc] init];
    self.embeddedViewContainer.translatesAutoresizingMaskIntoConstraints = NO;
    [stackView addArrangedSubview:self.embeddedViewContainer];

    [NSLayoutConstraint activateConstraints:@[
        [scrollView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
        [scrollView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [scrollView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [scrollView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],

        [stackView.topAnchor constraintEqualToAnchor:scrollView.topAnchor constant:16],
        [stackView.leadingAnchor constraintEqualToAnchor:scrollView.leadingAnchor constant:16],
        [stackView.trailingAnchor constraintEqualToAnchor:scrollView.trailingAnchor constant:-16],
        [stackView.bottomAnchor constraintEqualToAnchor:scrollView.bottomAnchor constant:-16],
        [stackView.widthAnchor constraintEqualToAnchor:scrollView.widthAnchor constant:-32],
    ]];
}

- (void)createEmbeddedView {
    __weak typeof(self) weakSelf = self;
    
    PLKEmbeddedLinkTokenConfiguration *configuration = [PLKEmbeddedLinkTokenConfiguration createWithToken:self.linkToken onSuccess:^(PLKLinkSuccess *success) {
        NSLog(@"public-token: %@ metadata: %@", success.publicToken, success.metadata.metadataJSON);
    }];
    
    configuration.onExit = ^(PLKLinkExit *exit) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) return;
        
        if (exit.error) {
            NSLog(@"exit with %@\n%@", exit.error, exit.metadata.metadataJSON);
            [strongSelf setState:LoadingStateError];
            NSString *displayMessage = exit.error.userInfo[kPLKExitErrorDisplayMessageKey];
            strongSelf.errorMessage = displayMessage ?: exit.error.localizedDescription ?: @"Exited with error";
            [strongSelf updateUI];
        } else {
            NSLog(@"exit with %@", exit.metadata.metadataJSON);
        }
    };
    
    configuration.onEvent = ^(PLKLinkEvent *event) {
        NSLog(@"Link Event: %@", event.eventName);
    };
    
    NSError *error = nil;
    UIView *embeddedView = [PLKPlaid createEmbeddedLinkUIViewWithConfiguration:configuration 
                                                                          error:&error 
                                                              presentationHandler:^(UIViewController *viewController) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (strongSelf) {
            [strongSelf presentViewController:viewController animated:YES completion:nil];
        }
    } 
                                                                dismissalHandler:^(UIViewController *viewController) {
        [viewController dismissViewControllerAnimated:YES completion:nil];
    }];
    
    if (error) {
        self.state = LoadingStateError;
        self.errorMessage = error.localizedDescription;
        [self updateUI];
    } else if (embeddedView) {
        self.embeddedSearchView = embeddedView;
        self.state = LoadingStateReady;
        [self addEmbeddedView];
        [self updateUI];
    }
}

- (void)addEmbeddedView {
    if (!self.embeddedSearchView) return;
    self.embeddedSearchView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.embeddedViewContainer addSubview:self.embeddedSearchView];

    [NSLayoutConstraint activateConstraints:@[
        [self.embeddedSearchView.topAnchor constraintEqualToAnchor:self.embeddedViewContainer.topAnchor],
        [self.embeddedSearchView.leadingAnchor constraintEqualToAnchor:self.embeddedViewContainer.leadingAnchor],
        [self.embeddedSearchView.trailingAnchor constraintEqualToAnchor:self.embeddedViewContainer.trailingAnchor],
        [self.embeddedSearchView.bottomAnchor constraintEqualToAnchor:self.embeddedViewContainer.bottomAnchor],
        
        // Set a minimum height for the embedded view
        [self.embeddedSearchView.heightAnchor constraintGreaterThanOrEqualToConstant:400]
    ]];
}

- (void)updateUI {
    dispatch_async(dispatch_get_main_queue(), ^{
        switch (self.state) {
            case LoadingStateLoading:
                self.errorLabel.hidden = YES;
                self.embeddedViewContainer.hidden = YES;
                break;
                
            case LoadingStateReady:
                self.errorLabel.hidden = YES;
                self.embeddedViewContainer.hidden = NO;
                break;
                
            case LoadingStateError:
                self.errorLabel.text = self.errorMessage;
                self.errorLabel.hidden = NO;
                self.embeddedViewContainer.hidden = YES;
                break;
        }
    });
}

@end

File: LinkDemo-UIKit/LinkDemo-UIKit/ViewControllers/PlaidEmbeddedLinkViewController.swift
//
//  PlaidEmbeddedLinkViewController.swift
//  LinkDemo-UIKit
//
//  Copyright © 2026 Plaid Inc. All rights reserved.
//

import LinkKit
import UIKit

class PlaidEmbeddedLinkViewController: UIViewController {

    init(linkToken: String) {
        self.linkToken = linkToken
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        title = "PlaidEmbeddedLink"
        view.backgroundColor = .systemBackground

        setupUI()
        updateUI()
        createEmbeddedView()
    }

    // MARK: - Link Setup

    private func createEmbeddedView() {
        let configuration = EmbeddedLinkTokenConfiguration(
            token: linkToken,
            onSuccess: { linkSuccess in
                print("public-token: \(linkSuccess.publicToken) metadata: \(linkSuccess.metadata)")
            },
            onExit: { [weak self] linkExit in
                if let error = linkExit.error {
                    print("exit with \(error)\n\(linkExit.metadata)")
                    DispatchQueue.main.async {
                        self?.state = .error(error.displayMessage ?? "Exited with error")
                    }
                } else {
                    print("exit with \(linkExit.metadata)")
                }
            },
            onEvent: { linkEvent in
                print("Link Event: \(linkEvent)")
            }
        )

        do {
            let embeddedView = try Plaid.createEmbeddedLinkUIView(
                configuration: configuration,
                presentationMethod: .viewController(self)
            )

            self.embeddedSearchView = embeddedView
            self.state = .ready

            embeddedView.translatesAutoresizingMaskIntoConstraints = false
            containerView.addSubview(embeddedView)

            NSLayoutConstraint.activate([
                embeddedView.topAnchor.constraint(equalTo: containerView.topAnchor),
                embeddedView.leadingAnchor.constraint(equalTo: containerView.leadingAnchor),
                embeddedView.trailingAnchor.constraint(equalTo: containerView.trailingAnchor),
                embeddedView.bottomAnchor.constraint(equalTo: containerView.bottomAnchor),
                embeddedView.heightAnchor.constraint(equalToConstant: 400),
            ])

        } catch {
            self.state = .error(error.localizedDescription)
        }
    }

    // MARK: - UI

    private let linkToken: String
    private var embeddedSearchView: UIView?
    private var state: LoadingState = .loading {
        didSet {
            updateUI()
        }
    }

    private let titleLabel = UILabel()
    private let versionLabel = UILabel()
    private let errorLabel = UILabel()
    private let containerView = UIView()

    private func updateUI() {
        switch state {
        case .loading:
            errorLabel.isHidden = true
        case .ready:
            errorLabel.isHidden = true
        case .error(let message):
            errorLabel.text = message
            errorLabel.isHidden = false
        }
    }

    private func setupUI() {
        let stackView = UIStackView()
        stackView.axis = .vertical
        stackView.spacing = 16
        stackView.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(stackView)

        titleLabel.text = "Plaid Embedded Link View Example"
        titleLabel.font = .boldSystemFont(ofSize: 20)
        titleLabel.textAlignment = .center
        titleLabel.numberOfLines = 0
        stackView.addArrangedSubview(titleLabel)

        versionLabel.text = "LinkKit"
        versionLabel.font = .systemFont(ofSize: 14)
        versionLabel.textColor = .secondaryLabel
        versionLabel.textAlignment = .center
        stackView.addArrangedSubview(versionLabel)

        errorLabel.font = .systemFont(ofSize: 14)
        errorLabel.textColor = .systemRed
        errorLabel.textAlignment = .center
        errorLabel.numberOfLines = 0
        errorLabel.isHidden = true
        stackView.addArrangedSubview(errorLabel)

        containerView.translatesAutoresizingMaskIntoConstraints = false
        stackView.addArrangedSubview(containerView)

        NSLayoutConstraint.activate([
            stackView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 16),
            stackView.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
            stackView.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
            stackView.bottomAnchor.constraint(lessThanOrEqualTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -16),
        ])
    }
}
